src/main/java/htsjdk/samtools/cram/encoding/readfeatures/Substitution.java

```diff
@@ -116,6 +117,11 @@ public boolean equals(final Object obj) {
         return true;
     }
 
+    @Override
+    public int hashCode() {
+        return Objects.hash(position, base, referenceBase, code);
```

Comment:
From an previous attempt to fix this: https://github.com/samtools/htsjdk/pull/1222#discussion_r231674367

Reply:
Specifically: what is that weird `NO_CODE` stuff in `equals()` and do we need to match its behavior here?

Reply:
From the [contract](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode--) for `hashCode`:

> If two objects are equal according to the `equals(Object)` method, then calling the `hashCode` method on each of the two objects must produce the same integer result.

If two `Substitution` objects are equal then they must have equal `position`, `base`, `referenceBase`, and `code` fields (since if any of them differ then `equals` returns false), and the implementation of `hashCode` will produce the same result for both objects.

Furthermore

> It is _not_ required that if two objects are unequal according to the `equals(java.lang.Object)` method, then calling the `hashCode` method on each of the two objects must produce distinct integer results.

In this case, if the two objects have `NO_CODE` codes then `equals` will return false, but `hashCode` may return `true` - which is consistent with the contract.

So, my reading is that this fix is OK.

Reply:
I think I've found a violation: when `a.code == b.code == NO_CODE` and `a.base != b.base`, then `equals()` returns true but `hashCode()`s would differ.

Reply:
@jmthibault79 you're right - and I am spectacularly wrong! I've fixed and added a comprehensive test.